src/file.py

```diff
@@ -0,0 +1,406 @@
+            )
+            raise AuthenticationException(common.ACCESS_ERROR)
+        if res.status.code == cs3code.CODE_NOT_FOUND:
+            self.log.info(f'msg="File not found on remove" filepath="{resource.file_ref}"')
+            raise IOError(common.ENOENT_MSG)
+        if res.status.code != cs3code.CODE_OK:
+            if "path not found" in str(res):
+                self.log.info(f'msg="Invoked removefile on non-existing file" filepath="{resource.file_ref}"')
+                raise IOError(common.ENOENT_MSG)
+            self.log.error(
+                'msg="Failed to remove file" filepath="%s" trace="%s" code="%s" reason="%s"'
+                % (resource.file_ref, res.status.trace, res.status.code, res.status.message.replace('"', "'"))
+            )
+            raise IOError(res.status.message)
+        self.log.debug(f'msg="Invoked removefile" result="{res}"')
+
+    def touch_file(self, resource: Resource) -> None:
+        """Touch a file using the given userid as access token."""
+        req = cs3sp.TouchFileRequest(
+            ref=resource.ref,
+            opaque=types.Opaque(map={"Upload-Length": types.OpaqueEntry(decoder="plain", value=str.encode("0"))}),
+        )
+        res = self.gateway.TouchFile(request=req, metadata=[("x-access-token", self.auth.token)])
+        if res.status.code == cs3code.CODE_FAILED_PRECONDITION:
+            self.log.info(
+                'msg="Lock mismatch touching file" filepath="%s" reason="%s"'
+                % (resource.file_ref, res.status.message.replace('"', "'"))
+            )
+            raise IOError(common.EXCL_ERROR)
+        if res.status.code == cs3code.CODE_UNAUTHENTICATED:
+            self.log.error(
+                'msg="Authentication failed on touch file" filepath="%s" userid="%s" trace="%s" reason="%s"'
+                % (resource.file_ref, self.auth.token[-20:], res.status.trace, res.status.message.replace('"', "'"))
+            )
+            raise AuthenticationException(common.ACCESS_ERROR)
+        if res.status.code != cs3code.CODE_OK:
+            self.log.error(
+                'msg="Failed to initiateFileUpload on touch" filepath="%s" trace="%s" code="%s" reason="%s"'
+                % (resource.file_ref, res.status.trace, res.status.code, res.status.message.replace('"', "'"))
+            )
+            raise IOError(res.status.message)
+        self.log.debug(f'msg="Invoked touchfile" result="{res}"')
+
+    def write_file(self, resource: Resource, content: Union[str, bytes], size: int, islock=False) -> None:
+        """Write a file using the given userid as access token. The entire content is written
+        and any pre-existing file is deleted (or moved to the previous version if supported).
+        The islock flag is currently not supported. The backend should at least support
+        writing the file with O_CREAT|O_EXCL flags to prevent races."""
+        tstart = time.time()
+        # prepare endpoint
+        if size == -1:
```

Comment:
I think it may be in order to create a file with size 0 instead of using "touch_file" which was not implemented in the wopi server, this would be redundant now.

Reply:
I'd need to check git history there, but IIRC this was added when implementing streaming, and `touch` would not work with Reva edge (i.e. OCIS)

Reply:
Okay maybe we should keep it then if Reva edge needs to be considered as well

Reply:
I will, yeah I don't think it serves any purpose.